File: 518-CoinChangeIi/518-CoinChangeIi.py
# Last updated: 2/23/2026, 6:46:40 PM
class Solution:
    def change(self, amount: int, coins: List[int]) -> int:
        # d[i][w] is the number of combinations with coins [0:i] and the amount is w.
        # Rules:
        # d[i][w] = d[i-1][w] + d[i][w - coins[i]]
        
        if amount == 0:
            return 1
        elif len(coins) == 0:
            return 0
        
        
        d = [[0 for _ in range(amount+1)] for __ in range(len(coins))]
        
        # Base: 
        for i in range(len(coins)):
            d[i][0] = 1
        
        for i in range(len(coins)):
            for w in range(1, amount+1):
                if w - coins[i] >=0:
                    d[i][w] = d[i-1][w] + d[i][w - coins[i]]
                else:
                    d[i][w] = d[i-1][w]
      
        # print(d)
        return d[-1][-1]

File: 518-CoinChangeIi/518-CoinChangeIi.java
// Last updated: 2/23/2026, 6:46:38 PM
class Solution {
    public int change(int amount, int[] coins) {
        if(amount == 0) {
            return 1;
        } else if (coins.length == 0) {
            return 0;
        }
        
        // Arrays.sort(coins);
        int[] b = new int[amount+1];
        b[0] = 1;
       
        // For one dimension space, we should always update
        // vertically to make sure that b[w] stores the value
        // for coins [1...i] before updating b[w].
        for(int i=0; i<coins.length; i++) { 
            for(int w=1; w<=amount; w++) {
                int v = coins[i];
                if(v > w) {
                    // so that b[w] still stores the value
                    // with coins [1...i-1]
                    continue;
                } 
                // b[w-v] stores the value for coin [1...i].
                b[w] += b[w - v]; 
            }
        }
        
        return b[amount];
    }
}